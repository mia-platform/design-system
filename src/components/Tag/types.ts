
export enum Type {
  Tag = 'Tag',
  Chip = 'Chip'
}

export enum Color {
  Grey = 'Grey',
  Blue = 'Blue',
  Teal = 'Teal',
  Green = 'Green',
  Yellow = 'Yellow',
  Magenta = 'Magenta',
  Purple = 'Purple',
  Red = 'Red',
}
