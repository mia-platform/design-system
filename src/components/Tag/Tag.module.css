.tag {
  margin-inline-end: 0px;
}
