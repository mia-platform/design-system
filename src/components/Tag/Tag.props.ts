/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactNode } from 'react'

import { Color, Type } from './types.ts'
import { IconComponent } from '../Icon/Icon.props.ts'

export type TagProps = {

    /**
     * The main content of the Tag.
     */
    children?: ReactNode,

    /**
    * The type of the Tag. Default 'Tag'.
    */
    type?: Type,

    /**
    * The type of the Tag. Default 'Tag'.
     */
    color?: Color | string,

    /**
    * Optional icon of the Tag.
    */
    icon?: IconComponent,

    /**
     * Displays the default close icon when set to true. It can also be used to render a custom icon.
     */
    closeIcon?: ReactNode,

    /**
     * Callback executed when Tag is closed. Only available if closeIcon is enabled.
     */
    onClose?: () => void,

    /**
     * Displays the Tag border. Defaults to true.
     */
    isBordered?: boolean,
}
