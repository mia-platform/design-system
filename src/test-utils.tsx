/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { RenderOptions, RenderResult, render } from '@testing-library/react'
import { ConfigProvider } from 'antd'
import { ReactNode } from 'react'

import { ThemeContext, defaultTheme } from './components/ThemeProvider'
import { ANT_PREFIX } from './components/ThemeProvider/Ant'

/**
 * Wraps a React node with a series of providers for testing purposes.
 *
 * @remarks It adds the default theme to components
 * @remarks It adds the default Ant class prefix
 * @remarks It removes the css-do-not-override classes from Ant Design components
 *
 * @param {ReactNode} children - The React node to be wrapped.
 * @returns {ReactNode} The wrapped React node.
 */
const TestWrapper = ({ children }: {children: ReactNode}): ReactNode => (
  <ThemeContext.Provider value={defaultTheme}>
    <ConfigProvider prefixCls={ANT_PREFIX} theme={{ hashed: false }}>
      {children}
    </ConfigProvider>
  </ThemeContext.Provider>
)

/**
 * Custom render function for testing React components.
 *
 * @param {ReactNode} element - The React node to be rendered.
 * @param {RenderOptions} [options] - Options for rendering.
 * @returns {RenderResult} The result of rendering the React node.
 */
const customRender = (
  element: ReactNode,
  options?: Omit<RenderOptions, 'wrapper'>,
): RenderResult => render(element, { wrapper: TestWrapper, ...options })

export * from '@testing-library/react'
export {
  customRender as render,
  TestWrapper,
}
