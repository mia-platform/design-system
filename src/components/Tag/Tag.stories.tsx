/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Meta, StoryObj } from '@storybook/react'

import { Tag } from '.'
import { defaults } from './Tag'

const meta = {
  component: Tag,
  args: {
    ...defaults,
    children: 'Tag text',
  },
  argTypes: {
    children: { control: false },
    closeIcon: { type: 'boolean' },
  },
} satisfies Meta<typeof Tag>

export default meta
type Story = StoryObj<typeof meta>

export const BasicExample: Story = {}

export const Colored: Story = {
  args: {
    color: 'magenta',
  },
}

export const Bordered: Story = {
  args: {
    isBordered: true,
  },
}

export const WithDefaultCloseIcon: Story = {
  args: {
    closeIcon: true,
  },
}

export const WithCustomCloseIcon: Story = {
  args: {
    closeIcon: <span>X</span>,
  },
}

