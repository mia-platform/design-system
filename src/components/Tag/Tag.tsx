/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactElement, useMemo } from 'react'
import { Tag as AntTag } from 'antd'
import classnames from 'classnames'

import { Color, Type } from './types.ts'
import { Icon } from '../Icon'
import { TagProps } from './Tag.props'
import styles from './Tag.module.css'

export const defaults = {
  isBordered: true,
  color: Color.Grey,
}

export const Tag = (
  {
    children,
    type,
    color = Color.Grey,
    closeIcon,
    onClose,
    icon,
    isBordered = defaults.isBordered,
  } : TagProps
) : ReactElement => {
  const className = useMemo(() => classnames([
    styles.tag,
    type === Type.Chip && styles.chip,
    color === Color.Grey && styles.tagColorGrey,
    color === Color.Blue && styles.tagColorBlue,
    color === Color.Teal && styles.tagColorTeal,
    color === Color.Green && styles.tagColorGreen,
    color === Color.Yellow && styles.tagColorYellow,
    color === Color.Magenta && styles.tagColorMagenta,
    color === Color.Purple && styles.tagColorPurple,
    color === Color.Red && styles.tagColorRed,
  ]), [color, type])

  const customColor = useMemo(() => {
    return color && !(color in Color) ? color : undefined
  }, [color])

  return (
    <AntTag
      bordered={isBordered}
      className={className}
      closeIcon={closeIcon}
      color={customColor}
      icon={icon && (
        <Icon component={icon} size={16} />
      )}
      onClose={onClose}
    >
      {children}
    </AntTag>
  )
}

Tag.Type = Type
Tag.Color = Color
