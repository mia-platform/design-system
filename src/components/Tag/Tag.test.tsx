/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { Tag } from './Tag'
import { render } from '../../test-utils'

const props = {
  color: '#ff0000',
  closeIcon: true,
  onClose: jest.fn(),
}

describe('Tag', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    const { asFragment } = render(<Tag {...props}>{'Tag text'}</Tag>)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly with close icon and calls onClose', async() => {
    render(<Tag {...props}>{'Tag text'}</Tag>)

    const closeButton = screen.getByRole('img', { name: /close/i })
    expect(closeButton).toBeInTheDocument()
    userEvent.click(closeButton)
    await waitFor(() => expect(props.onClose).toHaveBeenCalled())
  })
})
