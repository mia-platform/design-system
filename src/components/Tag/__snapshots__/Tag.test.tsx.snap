// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-tag mia-platform-tag-has-color"
    style="background-color: rgb(255, 0, 0);"
  >
    Tag text
    <span
      aria-label="close"
      class="anticon anticon-close mia-platform-tag-close-icon"
      role="img"
      tabindex="-1"
    >
      <svg
        aria-hidden="true"
        data-icon="close"
        fill="currentColor"
        fill-rule="evenodd"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
        />
      </svg>
    </span>
  </span>
</DocumentFragment>
`;
